### PR TITLE
Remove in-memory reader state

### LINKED ISSUE #6803: Remove caching of next Entry in MappedSegmentedReader
Remove caching of nextEntry in MappedSegmentedReader, instead hasNext can use FrameUtil to see if there is a next entry. 

We had to reset all readers when truncating because the in-memory state (nextEntry) can be corrupted. If we don't keep any in-memory state, only the readers which are at the truncated entries have to be reset. All concurrent readers are zeebe processing/exporting readers which only read committed entries. Since committed entries are never truncated, those readers never have to be reset as long as we don't cache the next entry. This would help us to alleviate some concurrency issues between truncation and readers #4198

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

This PR has 3 commits which adds the following changes:
* Remove keeping track of `previousEntry` in SegmentedJournalReader. It was never used and incorrectly updated.
* Remove caching nextEntry in MappedSegmentedReader, instead use the frame to check if next entry exists.
* Since we use frame for `hasNext()`, it is important that the nextEntry is never null. Hence the reader util thrown exception if it is not able to read a valid record.

## Related issues

closes #6803 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: journal/src/main/java/io/zeebe/journal/file/FrameUtil.java [hunk 1]
```diff
@@ -49,14 +49,22 @@ public static void markAsIgnored(final ByteBuffer buffer, final int offset) {
...
     return Optional.empty();
   }
 
+  public static boolean hasValidVersion(final ByteBuffer buffer, final int position) {
+    if (buffer.capacity() < position + LENGTH) {
+      return false;
+    }
+    final byte val = buffer.get(position);
+    return val != IGNORE;
+  }
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/FrameUtil.java:66]: Why not use the existing code that already reads the version?

```suggestion
  public static boolean hasValidVersion(final ByteBuffer buffer, final int position) {
    final boolean valid = readVersion(buffer).isPresent();
    buffer.reset();
    return valid;
  }
```

File: journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentReader.java [hunk 2]
```diff
@@ -44,89 +43,53 @@
...
-    nextEntry = null;
-
-    // Read the next entry in the segment.
-    readNext(getNextIndex());
-
-    // Return the current entry.
+    final Optional<Integer> version = FrameUtil.readVersion(buffer);
+    if (version.isEmpty()) {
+      return null;
+    }
```
Comment (MiguelPires) [inline journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentReader.java:59]: If we already checked the validity of the entry then we don't need this check. Maybe the optional no longer makes sense given the new usage

**Review by miguelpires — CHANGES_REQUESTED** [review]
A few minor comments. 

I think that the Optional might not make sense anymore. It was a way of reading and validating the frame in one operation but if the current usage needs to separate them then it might be simpler to just remove it

File: journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentReader.java (hunk 2 above)
Comment (deepthidevaki) [reply journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentReader.java:59]: Agree. I will remove the optional, and make sure we call `hasVersion` before every `readVersion`.

**Comment by deepthidevaki** [comment]
@MiguelPires I have applied your comments.

**Review by miguelpires — APPROVED** [review]
Thanks! :+1:

**Pull request merged**